### ISSUE TITLE
get_moreDA_augmentations segmentation resampling order parameter should be 0 (order_seg=0)

### ISSUE THREAD

**Issue by silvandeleemput** [MIC-DKFZ/nnUNet#909]
The default `nnUNetTrainerV2` trainer uses the  `get_moreDA_augmentation` method to setup the data generators.

https://github.com/MIC-DKFZ/nnUNet/blob/96d44c2fc1ce5e18da4cd54bf52882047c37982e/nnunet/training/data_augmentation/data_augmentation_moreDA.py#L41-L46

However, the `get_moreDA_augmentation` method has the parameter for the resampling order for the segmentation masks set to `order_seg=1`. Shouldn't this be set to `order_seg=0` to use nearest neighbors resampling, which is typically used for segmentation mask resampling?

**Comment by FabianIsensee** [comment]
nah order=1 is more accurate in my experience. There is no reason this has to be 0.
Note though that I don't really see a difference in segmentation performance between the two so you could also set it to 0 if you really desire (this would also reduce CPU utilization)

**Comment by silvandeleemput** [comment]
You might want to consider some potentially harmful side-effects by setting this to order=1. At the borders of resampled segmentations, you risk getting strange class values of classes that don't belong there. Take for example the dummy image below with three classes (values 1, 3, and 9) bordering each other, by doubling the image size with order=1, you can get these class artifacts, whereas by resampling with image order=0 this doesn't happen. Similar artifacts can be obtained by downscaling the image at a lower resolution.

![image](https://user-images.githubusercontent.com/3715102/149409511-d83b35b3-bf92-4903-bfd6-7acf4abf0b48.png)

I would think that the segmentation order of 0 is the better default to avoid potential unwanted behavior.

**Comment by FabianIsensee** [comment]
This is long taken care of ;-) If you follow the code you will see it.
Segmentations are resampled by first converting them to a one-hot encoding. Then each encoding is resampled separately and finally the results are converted back to a segmentation map. No strange artifacts in nnU-net.

**Comment by silvandeleemput** [comment]
I agree that when you use one-hot encodings for the segmentations and resample them per channel this could make sense with resampling order=1. However, I went through the entire data augmentation pipeline (unmodified vanilla `nnUNetTrainerV2` trainer) (using the converted `Task004_Hippocampus` data/task), but I couldn't find any code/operation converting the segmentations to one-hot-encodings.

Here is a debug dump after each transform in the `get_moreDA_augmentation` training data generator, together with the shape of the segmentation data and the unique data values:

```
 0 Applying transform: SegChannelSelectionTransform             seg    shape: (9, 1, 73, 80, 64) unique vals: [-1.  0.  1.  2.]
 1 Applying transform: SpatialTransformWithWeights              seg    shape: (9, 1, 40, 56, 40) unique vals: [-1.  0.  1.  2.]
 2 Applying transform: GaussianNoiseTransform                   seg    shape: (9, 1, 40, 56, 40) unique vals: [-1.  0.  1.  2.]
 3 Applying transform: GaussianBlurTransform                    seg    shape: (9, 1, 40, 56, 40) unique vals: [-1.  0.  1.  2.]
 4 Applying transform: BrightnessMultiplicativeTransform        seg    shape: (9, 1, 40, 56, 40) unique vals: [-1.  0.  1.  2.]
 5 Applying transform: ContrastAugmentationTransform            seg    shape: (9, 1, 40, 56, 40) unique vals: [-1.  0.  1.  2.]
 6 Applying transform: SimulateLowResolutionTransform           seg    shape: (9, 1, 40, 56, 40) unique vals: [-1.  0.  1.  2.]
 7 Applying transform: GammaTransform                           seg    shape: (9, 1, 40, 56, 40) unique vals: [-1.  0.  1.  2.]
 8 Applying transform: GammaTransform                           seg    shape: (9, 1, 40, 56, 40) unique vals: [-1.  0.  1.  2.]
 9 Applying transform: MirrorTransformWithWeights               seg    shape: (9, 1, 40, 56, 40) unique vals: [-1.  0.  1.  2.]
10 Applying transform: MaskTransform                            seg    shape: (9, 1, 40, 56, 40) unique vals: [-1.  0.  1.  2.]
11 Applying transform: RemoveLabelTransform                     seg    shape: (9, 1, 40, 56, 40) unique vals: [0. 1. 2.]
12 Applying transform: RenameTransform                          target shape: (9, 1, 40, 56, 40) unique vals: [0. 1. 2.]
...
```

The segmentations after the preprocessing before the augmentations that are loaded by the dataloader are segmentations maps that don't seem to be one-hot encoded (single channel and data values > 1). This doesn't seem to change anywhere in the augmentation pipeline. Also, the CrossEntropy and Dice losses seem to take the `target` as a normal segmentation map and not as a one-hot encoded image.

To me it seems that there is no one-hot-encoding being used during training for the segmentations/targets using the default `nnUNetTrainerV2` (no operation that makes the seg/targets have multiple channels matching the number of classes with (0, 1 or similar) encodings). Is this correct? or could you point me to the place where the one-hot encoding is applied in case I missed it?

**Comment by FabianIsensee** [comment]
From all those transforms only SpatialTransformWithWeights and SimulateLowResolutionTransform use interpolation and in there is the code you are looking for. The one hot encoding is only done temporarily during resampling. This is why you don't see it

**Comment by FabianIsensee** [comment]
https://github.com/MIC-DKFZ/batchgenerators/blob/10c47ddbca8ccce1e575b912615e325fb42b74e8/batchgenerators/augmentations/utils.py#L583

**Comment by FabianIsensee** [comment]
https://github.com/MIC-DKFZ/batchgenerators/blob/10c47ddbca8ccce1e575b912615e325fb42b74e8/batchgenerators/augmentations/utils.py#L144

**Comment by silvandeleemput** [comment]
Ah, I totally missed that, thanks for pointing it out!
